### ISSUE TITLE
Check and verify a specific snapshot

### ISSUE THREAD

**Issue by akrabu** [restic/restic#3326]
Output of `restic version`
--------------------------

restic 0.12.0 compiled with go1.15.7 on darwin/amd64

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------

restic check --read-data --snapshot 9170996b
^ Would verify all data referenced by a specific snapshot

restic check --read-data-subset --snapshot 9a70996b
^ Would verify just the data added in that particular snapshot

What are you trying to do? What problem would this solve?
---------------------------------------------------------

It would be handy for testing purposes to ensure that a specific snapshot was readable.  For instance, if you had a connection hiccup in the middle of a specific backup and knew it, you could do a --read-data-subset --snapshot 1234abcd and verify that the blobs uploaded just during that snapshot were intact.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------

Always!  <3

**Comment by aawsome** [comment]
I like the idea! If I find some time, I can try to work on a PR!

I would however change the syntax to 
```
restic check [flags] [snapshot ID] [...]
```
and let users specify the snapshot(s) like in other commends, e.g. `forget`. That would also allow to do a `restic check --read-data latest`.

What we should discuss:

- Should we allow snapshots only in combination with `--read-data*`?
- If not, how to handle non-packfile-errors in the repository which would not affect these?

**Comment by akrabu** [comment]
Hmm that's a good question.  I'll defer to those who know better than me, but I could see where having the same options as, say, "find" might make sense.  So, if I have a corrupted database, and I do a --read-data and find the pack or tree affected, then do a rebuild-index and a fresh backup...  might it make sense to be able to read JUST that pack / tree?  Or does rebuild-index rename all of those packs / trees?  I'm not 100% clear on that.  If they stay the same, it might make sense.  If they don't, it might not haha

**Comment by akrabu** [comment]
I just thought of a use-case outside of "snapshots" alone.  What about when you do a --read-data and have a connection hiccup, and you want to go back over JUST the data that didn't read properly during the hiccup?  For instance I have:

```
 rclone: 2021/03/09 00:25:11 ERROR : data/b7/b71340505ffd6cae4fcc9fb57c9ab6ed7a2450c912582636627ea0a29ce52d89: Didn't finish writing GET request (wrote 262144/4972912 bytes): unexpected EOF
Load(<data/b71340505f>, 0, 0) returned error, retrying after 720.254544ms: unexpected EOF
rclone: 2021/03/09 06:18:13 ERROR : data/72/72ffb201d81bbeefe54cb3fd94daa30172ce647b51bfe24935dba5705bbd8600: Didn't finish writing GET request (wrote 262144/4979353 bytes): unexpected EOF
Load(<data/72ffb201d8>, 0, 0) returned error, retrying after 582.280027ms: unexpected EOF
rclone: 2021/03/09 06:18:59 ERROR : data/f4/f4842c0fc6ce960bab05ce2fba058a9dcafe470623a1520da19cedfc7416dd60: Didn't finish writing GET request (wrote 786432/4667940 bytes): unexpected EOF
Load(<data/f4842c0fc6>, 0, 0) returned error, retrying after 468.857094ms: unexpected EOF

```

It would be nice to be able to check those specific blobs, and not have to do an entire --read-data again, just to ensure they're OK.

**Comment by aawsome** [comment]
> It would be nice to be able to check those specific blobs, and not have to do an entire --read-data again, just to ensure they're OK.

This can be handled by #3203

**Comment by MichaelEischer** [comment]
> That would also allow to do a `restic check --read-data latest`.

What would that command check? Everything referenced by the latest snapshot or just what was added in that snapshot (that latter variant is not really implementable)? I also wonder how much that would speed up the check process? When would one use that instead of just running `check`?

>     * Should we allow snapshots only in combination with `--read-data*`?

If you just want to check the tree metadata of the latest snapshot, then there's not much reason to enforce that combination.

>     * If not, how to handle non-packfile-errors in the repository which would not affect these?

You mean errors reported for snapshots which were not specified on the command line?

> So, if I have a corrupted database, and I do a --read-data and find the pack or tree affected, then do a rebuild-index and a fresh backup... it might make sense to be able to read JUST that pack / tree (for both time and cost reasons)?

The full `--read-data` check should only be necessary once. Afterwards just remove/salvage those pack files or use #3203 to recheck those. After running rebuild-index a plain run of `check` is enough.

> Or does rebuild-index rename all of those packs / trees? I just know the snapshot IDs get renamed.

`rebuild-index` does not change any blob/tree/snapshot IDs.

**Comment by rawtaz** [comment]
>> That would also allow to do a `restic check --read-data latest`.

>What would that command check? Everything referenced by the latest snapshot or just what was added in that snapshot (that latter variant is not really implementable)?

It makes little sense to check only part of a snapshot (e.g. the incremental changes in it), because:

- The check command is still about checking that your data can actually be restored, and you never restore "the parts of a snapshot that were changed" except if you know exactly which files or parts of files that were changed in that snapshot, in which case we're really talking about a much more fine-grained type of check that should check *files*, not parts of snapshots.

- If you were to check only the changed data in a snapshot, you'd still have to check some or all parts of parent snapshots, to know that this specific snapshot can be restored entirely.

In summary, `restic check --read-data [snapshot ID]` should check all of the data referenced by that snapshot.

**Comment by hvhaugwitz** [comment]
What is the state of this features request?
